xds/src/test/java/io/grpc/xds/XdsSdsClientServerTest.java

```diff
@@ -0,0 +1,224 @@
+  private Server server;
+
+  @Test
+  public void plaintextClientServer() throws IOException {
+    getXdsServer(/* downstreamTlsContext= */ null);
+    buildClientAndTest(
+        /* upstreamTlsContext= */ null, /* overrideAuthority= */ null, "buddy", server.getPort());
+  }
+
+  /** TLS channel - no mTLS. */
+  @Test
+  public void tlsClientServer_noClientAuthentication() throws IOException {
+    DownstreamTlsContext downstreamTlsContext =
+        SecretVolumeSslContextProviderTest.buildDownstreamTlsContextFromFilenames(
+            SecretVolumeSslContextProviderTest.SERVER_1_KEY_FILE,
+            SecretVolumeSslContextProviderTest.SERVER_1_PEM_FILE,
+            null);
+
+    getXdsServer(downstreamTlsContext);
+
+    // for TLS, client only needs trustCa
+    UpstreamTlsContext upstreamTlsContext =
+        SecretVolumeSslContextProviderTest.buildUpstreamTlsContextFromFilenames(
+            null, null, SecretVolumeSslContextProviderTest.CA_PEM_FILE);
+    buildClientAndTest(upstreamTlsContext, "foo.test.google.fr", "buddy", server.getPort());
+  }
+
+  /** mTLS - client auth enabled. */
+  @Test
+  public void mtlsClientServer_withClientAuthentication() throws IOException {
+    UpstreamTlsContext upstreamTlsContext =
+        SecretVolumeSslContextProviderTest.buildUpstreamTlsContextFromFilenames(
+            SecretVolumeSslContextProviderTest.CLIENT_KEY_FILE,
+            SecretVolumeSslContextProviderTest.CLIENT_PEM_FILE,
+            SecretVolumeSslContextProviderTest.CA_PEM_FILE);
+    mtlsCommonTest(upstreamTlsContext);
+  }
+
+  /** mTLS - client auth enabled then update server certs to untrusted. */
+  @Test
+  public void mtlsClientServer_changeServerContext_expectException() throws IOException {
+    UpstreamTlsContext upstreamTlsContext =
+        SecretVolumeSslContextProviderTest.buildUpstreamTlsContextFromFilenames(
+            SecretVolumeSslContextProviderTest.CLIENT_KEY_FILE,
+            SecretVolumeSslContextProviderTest.CLIENT_PEM_FILE,
+            SecretVolumeSslContextProviderTest.CA_PEM_FILE);
+    XdsClient.ListenerWatcher listenerWatcher = mtlsCommonTest(upstreamTlsContext);
+    DownstreamTlsContext downstreamTlsContext =
+        SecretVolumeSslContextProviderTest.buildDownstreamTlsContextFromFilenames(
+            BAD_SERVER_KEY_FILE,
+            BAD_SERVER_PEM_FILE,
+            SecretVolumeSslContextProviderTest.CA_PEM_FILE);
+    createListenerUpdate(server.getPort(), downstreamTlsContext, listenerWatcher);
+    try {
+      buildClientAndTest(upstreamTlsContext, "foo.test.google.fr", "buddy", server.getPort());
+      fail("exception expected");
+    } catch (StatusRuntimeException sre) {
+      assertThat(sre).hasCauseThat().isInstanceOf(SSLHandshakeException.class);
+      assertThat(sre).hasCauseThat().hasMessageThat().isEqualTo("General OpenSslEngine problem");
+    }
+  }
+
+  private XdsClient.ListenerWatcher mtlsCommonTest(UpstreamTlsContext upstreamTlsContext)
```

Comment:
why test returns something unrelated?

Reply:
because it is used in `mtlsClientServer_changeServerContext_expectException`. I can change it from a return value to a class instance variable or rename it to address this comment

Reply:
renamed the method which also addresses the other comment about method name